### PR TITLE
ENH: Added AE as reduction tool

### PR DESCRIPTION
**Pull request by giovastabile**

### CODE REVIEW

File: ezyrb/ae.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
+"""
+Module for FNN-Autoencoders.
+"""
+import torch
+import torch.nn as nn
+from .reduction import Reduction
+from .ann import ANN
+
+class AE(ANN,Reduction):
```
Comment (mtezzele) [inline ezyrb/ae.py:9]: space after comma

File: ezyrb/ae.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    """
+    Feed-Forward AutoEncoder class (AE)
```
Comment (mtezzele) [inline ezyrb/ae.py:11]: check the doc generation, I think you need a blank line before :param

File: ezyrb/ae.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        self.function_decoder = function_decoder
+        self.loss = loss
+
+        self.stop_training = stop_training
+        self.loss_trend = []
+        self.encoder = None
+        self.decoder = None
+        self.optimizer = None
+        self.model = None
+    class innerAE(torch.nn.Module):
```
Comment (mtezzele) [inline ezyrb/ae.py:53]: Blank line and naming convention

File: ezyrb/ae.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        """
+        Autoencoder as a pytorch module
+        """
+        def __init__(self, outer_instance):
+            super().__init__()
+            self.encoder = outer_instance.encoder
+            self.decoder = outer_instance.decoder
+        def forward(self, x):
```
Comment (mtezzele) [inline ezyrb/ae.py:61]: blank line

File: ezyrb/ae.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            return decoded
+
+
+    def _build_model(self, values):
+        """
+        Build the torch model.
+
+        Considering the number of neurons per layer (self.layers), a
+        feed-forward NN is defined:
+            - activation function from layer i>=0 to layer i+1:
```
Comment (mtezzele) [inline ezyrb/ae.py:76]: check whether the list is properly formatted in the doc

File: ezyrb/ae.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            y_pred = self.model(values)
+            loss = self.loss(y_pred, values)
+            self.optimizer.zero_grad()
+            loss.backward()
+            self.optimizer.step()
+            self.loss_trend.append(loss.item())
+            print(n_epoch)
+            for criteria in self.stop_training:
+                if isinstance(criteria, int):
+                # stop criteria is an integer
```
Comment (mtezzele) [inline ezyrb/ae.py:142]: fix indentation of the comments

File: ezyrb/ae.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                    if n_epoch == criteria:
+                        flag = True
+                elif isinstance(criteria, float):
+                   # stop criteria is float
```
Comment (mtezzele) [inline ezyrb/ae.py:146]: fix indentation of the comments

File: ezyrb/ae.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                    if loss.item() < criteria:
+                        flag = True
+
+            if flag:
```
Comment (mtezzele) [inline ezyrb/ae.py:150]: is it necessary? I suppose the break within the previous loop will do the job

File: ezyrb/ae.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        """
+        X = self._convert_numpy_to_torch(X)
+        g = self.encoder(X)
+        return g.cpu().detach().numpy().T
+
+    def expand(self, g):
+        """
+        Projects a reduced to full order solution.
+
+        :type: numpy.ndarray
```
Comment (mtezzele) [inline ezyrb/ae.py:170]: add the :param keyword

**Review by ndem0 — APPROVED** [review]

**Pull request merged**